samtranslator/model/sam_resources.py

```diff
@@ -2297,11 +2297,12 @@ def _construct_ddb_datasources(
         resources: List[Resource] = []
 
         for relative_id, ddb_datasource in ddb_datasources.items():
+            datasource_logical_id = self.logical_id + relative_id
             cfn_datasource = DataSource(
-                logical_id=relative_id, depends_on=self.depends_on, attributes=self.resource_attributes
+                logical_id=datasource_logical_id, depends_on=self.depends_on, attributes=self.resource_attributes
             )
 
-            cfn_datasource.Name = ddb_datasource.Name or relative_id
+            cfn_datasource.Name = ddb_datasource.Name or datasource_logical_id
```

Comment:
Why also change the `Name`? Must it be unique within a region?

Reply:
Pretty sure, or else customers wouldn't be able to reference by `DataSourceName` in Resolver/Function.

Reply:
`DataSourceName` must be unique __only__ within GraphQL API

Reply:
So wouldn't it be better to keep it as-is? The "relative ID" is way clearer as it's verbatim what a customer specified.

Reply:
Yes, adding Resource prefix doesn't help being unique within resource

Reply:
<img width="845" alt="image" src="https://user-images.githubusercontent.com/67450797/225435105-a4301272-b351-4619-9e97-b8fbffc88a37.png">

Customer will get this error when deploying if they have two `AppSync::DataSource` with the same `Name` property.

Reply:
Which makes sense, how would the reference property `DataSourceName` work if you couldn't uniquely identify the `DataSource` resource by its name?

Reply:
Thanks, can you add a comment above `Name`? That's important context that cannot be deduced from code alone.

Reply:
Let's say customer creates `SameName` DS 2 times inside SAM resource - customer's screwed, if they create 2 `SameName` DS outside SAM resource - they'll get the error you posted. If they create 1 `SameName` in SAM and 1 outside - they'll be protected by your current change, but they'll create `SamSameName` and `SameName` DS which seems to be a backdoor within a rule "Make your DS names unique" and doesn't solve the issue. That's why I agree with Chris, that we don't need to add Resource prefix to the generated Name

Reply:
My line of thinking is this:

Customers will likely want the name of the DataSource to be the same as the relative id, so they won't define it and it will get generated by us with this prefix. This avoids conflicts if there are two `Serverless::GraphQLApi` with the same `DataSource` relative ID.

The case you talk about with the 1 `SameName` DataSource in SAM and 1 outside feels like a much more rare case to me that should not be optimized around IMO. And I also don't see the problem with us avoiding this name conflict here, even if it is inadvertent.

Reply:
Case with 2 SAM GraphQLApi in the same template, with the same name also seems to be rarely possible.

And yes customers likely will rely on relative_id for name. And the more I think about it, it feels wrong to me to change Name as an internal attribute of the abstraction. Generating logical ID for CFN resource is one thing. There was no that resource before, that logical ID is purely technical. Name, on the other hand, does exist. Even if it's not created explicitly, customer can think of it as of relevant_id. And then if we introduce another name it might even have some implications (ie if customer uses it somewhere in their code). SAM doesn't guarantee anything about logical IDs but I don't think it's expected by customers that we start changing names

Reply:
Ok I think I'm starting to see the concern here. You're right, it's not great for us to change `Name` property beyond the relative id, since the customer sets that value and they can manipulate it as such and would probably want full control over setting it.

I'm inclined to move ahead with leaving `Name` property as relative id. Are we ok with foregoing these other cases of duplicate names across numerous `Serverless::GraphQLApi` resources? The good thing is that the error message is clear and customers can immediately identify the problem. And as you said, this use case is also rare so it would not be something a customer encounters often at all.

Reply:
yep, lets keep name unchanged